btree/src/storage.rs

```diff
@@ -35,62 +25,74 @@ impl MmapStorage {
+            return Err(location + count);
+        }
+
+        // the file_len is only used in the destructor, to make the file's size on disk be the right one,
+        // at that point there is only one instance, so race conditions originated from this are unlikely
+        if location + count > self.file_len.load(Ordering::Acquire) {
+            self.file_len.store(location + count, Ordering::Release);
+        }
 
-    fn get(&'a self, location: u64, count: u64) -> Result<Self::Output, io::Error> {
         let location: usize = location.try_into().unwrap();
         let count: usize = count.try_into().unwrap();
-        Ok(&self.mmap[location..location + count])
+
+        // this unwrap can't fail because we already checked for size before
+        // I don't think we need any extra synchronization here,
+        // at least I don't think get_mut modifies any shared state
+
+        Ok((*self.mmap.get())
+            .get_mut(location..location + count)
+            .unwrap())
     }
 
-    fn put(&mut self, location: u64, bytes: impl AsRef<[u8]>) -> Result<(), std::io::Error> {
-        if location + u64::try_from(bytes.as_ref().len()).unwrap() > self.allocated_size {
-            self.allocated_size = Self::next_power_of_two(
-                self.allocated_size + location + u64::try_from(bytes.as_ref().len()).unwrap(),
-            );
+    pub fn extend(&mut self, minimum_required_size: u64) -> Result<(), io::Error> {
+        if minimum_required_size > self.allocated_size {
+            self.allocated_size = Self::next_power_of_two(minimum_required_size);
 
             // we need to extend the file, so we unmap, extend and remap
             unsafe {
-                self.mmap.flush()?;
+                // it's really important to flush here
+                self.sync()?;
                 ManuallyDrop::drop(&mut self.mmap);
                 (&mut *self.file).set_len(self.allocated_size)?;
-                self.mmap = ManuallyDrop::new(MmapMut::map_mut(&*self.file)?);
+                self.mmap = ManuallyDrop::new(UnsafeCell::new(MmapMut::map_mut(&*self.file)?));
             }
         }
-
-        if location + u64::try_from(bytes.as_ref().len()).unwrap() > self.file_len {
-            self.file_len = location + u64::try_from(bytes.as_ref().len()).unwrap();
-        }
-
-        let location: usize = location.try_into().unwrap();
-        self.mmap
-            .get_mut(location..location + bytes.as_ref().len())
-            .unwrap()
-            .copy_from_slice(&bytes.as_ref());
         Ok(())
     }
 
-    fn sync(&self) -> Result<(), io::Error> {
-        self.mmap.flush()
+    pub fn sync(&self) -> Result<(), io::Error> {
+        // there is nothing really unsafe here, we need the block only because of unsafe cell (at least nothing that is not already present in the memmap api)
+        unsafe { &*self.mmap.get() }.flush()
+    }
+
+    fn next_power_of_two(n: u64) -> u64 {
```

Comment:
This is just:

https://doc.rust-lang.org/1.0.0/core/primitive.u64.html#method.next_power_of_two